From 98fda6792e8471c64fd3d28438121c2d0adcd870 Mon Sep 17 00:00:00 2001
From: Drew Moseley <drew.moseley@northern.tech>
Date: Thu, 4 Jan 2018 11:35:39 -0500
Subject: [PATCH] vexpress_ca9x4: enable booting from ubifs

Enable options that add support for UBI images on CFI and UBIFS.

Enable MTD concatenation, thus 2*64MB NOR flash appears as single 128MB device.

Define the follwing flash partitions:
 - 1MB - uboot
    - 1MB - uboot environment (previously stored at the top of the flash)
          - remaining - UBI

When defining mtdparts use device name that matches kernel's MTD name (otherwise
partitioning will not be applied).

Signed-off-by: Maciej Borzecki <maciej.borzecki@rndity.com>

Backported to morty

Signed-off-by: Drew Moseley <drew.moseley@northern.tech>
---
 configs/vexpress_ca9x4_defconfig |  1 +
 include/configs/vexpress_ca9x4.h | 59 ++++++++++++++++++++++++++++++++++++++++
 2 files changed, 60 insertions(+)

diff --git a/configs/vexpress_ca9x4_defconfig b/configs/vexpress_ca9x4_defconfig
index 2947fc1..104b098 100644
--- a/configs/vexpress_ca9x4_defconfig
+++ b/configs/vexpress_ca9x4_defconfig
@@ -14,3 +14,4 @@ CONFIG_TARGET_VEXPRESS_CA9X4=y
 # CONFIG_CMD_SETEXPR is not set
 # CONFIG_CMD_NFS is not set
 # CONFIG_CMD_MISC is not set
+CONFIG_CMD_UBI=y
\ No newline at end of file
diff --git a/include/configs/vexpress_ca9x4.h b/include/configs/vexpress_ca9x4.h
index 99be50a..99f1135 100644
--- a/include/configs/vexpress_ca9x4.h
+++ b/include/configs/vexpress_ca9x4.h
@@ -15,4 +15,63 @@
 #define CONFIG_BOOTP_VCI_STRING     "U-Boot.armv7.vexpress_ca9x4"
 #include "vexpress_common.h"
 
+#define CONFIG_RBTREE
+#define CONFIG_CMD_UBI
+#define CONFIG_CMD_UBIFS
+#define CONFIG_CMD_MTDPARTS
+#define CONFIG_MTD_DEVICE
+#define CONFIG_MTD_PARTITIONS
+#define CONFIG_MTD_CONCAT
+#define CONFIG_FLASH_CFI_MTD
+#define CONFIG_LZO
+
+#ifdef CONFIG_SYS_MALLOC_LEN
+#undef CONFIG_SYS_MALLOC_LEN
+#define CONFIG_SYS_MALLOC_LEN (8 * 1024 * 1024)
+#endif
+
+/* CONFIG_MTD_CONCAT is set, hence nor0 and nor1 will become a single device
+   named nor2
+
+   NOTE: NOR flash appears as a single MTD device with name 40000000.flash */
+#define MTDIDS_DEFAULT		"nor2=40000000.flash"
+#define MTDPARTS_DEFAULT  "mtdparts=40000000.flash:"  \
+  "1m(u-boot)ro," \
+  "1m(u-boot-env)ro," \
+  "-(ubi)"
+
+#ifdef CONFIG_EXTRA_ENV_SETTINGS
+#undef CONFIG_EXTRA_ENV_SETTINGS
+#endif
+
+#define UBI_BOOTCMD \
+ 	"ubiboot=" \
+	"echo Booting from NOR...; "                                          \
+	"ubi part ubi && "                                                    \
+	"ubifsmount ubi0:rootfs && "                                          \
+	"ubifsload ${kernel_addr_r} /boot/zImage && "                         \
+	"ubifsload ${fdt_addr_r} /boot/${fdtfile} && "                        \
+  "setenv bootargs ${mtdparts} ${ubiargs} ${defargs} && "               \
+	"bootz ${kernel_addr_r} - ${fdt_addr_r}\0"
+
+#define CONFIG_EXTRA_ENV_SETTINGS                                       \
+  UBI_BOOTCMD \
+  "kernel_addr_r=0x60100000\0"                                          \
+  "fdt_addr_r=0x60000000\0"                                             \
+  "fdtfile=vexpress-v2p-ca9.dtb\0"                                      \
+  "mtdparts=" MTDPARTS_DEFAULT "\0"                                     \
+  "ubiargs=ubi.mtd=2 root=ubi0:rootfs rootfstype=ubifs ubi.fm_autoconvert=1\0" \
+  "defargs=console=ttyAMA0,115200n8 panic=3\0"
+
+#undef CONFIG_BOOTCOMMAND
+#define CONFIG_BOOTCOMMAND "run ubiboot; reset; "
+
+/* environment starts at 1MB offset in the flash */
+#undef CONFIG_ENV_OFFSET
+#define CONFIG_ENV_OFFSET 0x100000
+/* was fixed in vexpress_common, but will be derived in environment.h */
+#undef CONFIG_ENV_ADDR
+/* required for automatic calculation of CONFIG_ENV_ADDR */
+#define CONFIG_SYS_FLASH_BASE CONFIG_SYS_FLASH_BASE0
+
 #endif /* VEXPRESS_CA9X4_H */
-- 
2.7.4

